### ISSUE TITLE
grpc worker->client fetch tops out at 120 MB/s

### ISSUE THREAD

**Issue by yaroslavvb** [tensorflow/tensorflow#6256]
Fetching data as numpy arrays is slow for grpc sessions. Evaluating 128MB variable repeatedly from `Session("grpc://localhost..")` only gets me 116 MB/s, compared to 5150MB/s for `Session()`

https://gist.github.com/yaroslavvb/59bc8d3d635e1027306486cc418b26aa
```
python client_transfer_benchmark.py --profile
5150.04 MB per second
116.70 MB per second

```
Attached are CPU profiles for worker and for client. Worker uses 2x CPU and both of them are spending majority of the time in in gRPC code, specifically doing `strcmp` (reported as [__nss_hosts_lookup](https://github.com/gperftools/gperftools/issues/1)). 

[profile.client.pdf](https://github.com/tensorflow/tensorflow/files/644905/profile.client.pdf)
[profile.worker.pdf](https://github.com/tensorflow/tensorflow/files/644906/profile.worker.pdf)

This is related but slightly different from https://github.com/tensorflow/tensorflow/issues/6116 -- that issue looks at worker->worker transfer which hits 520 MB/second for the same message size (degrading below 100 MB/second for larger messages)

https://gist.github.com/yaroslavvb/e196107b5e0afc834652bd3153030c42
```
python benchmark_grpc_recv.py --data_mb=128
Local rate:       15214.63 MB/s
Distributed rate: 569.27 MB/s


```

**Comment by michaelisard** [comment]
@mrry do you want to comment on potential fixes in the future, or workarounds for now?

**Comment by mrry** [comment]
We're aware of the inefficiencies in the client-to-and-from-master path when using the distributed runtime, and have plans to cut down on the serialization overhead, especially in the common case where the client and master (and often worker) are colocated in the same process.

@yaroslavvb Can you share more about what your use case is for fetching large values from the distributed runtime? We haven't spent much time optimizing this until now because the common case in distributed execution involves running a training op that returns nothing (or perhaps some small progress indicator values). It would be good to understand your use case so we can prioritize the various steps we can take to improve performance here!

**Comment by shiyemin** [comment]
@mrry Hi, I'm also using distributed training on one server because it's easier to implement and extend. 

When using a large network (e.g. Inception-resnet-2) or batchsize, the traffic between session will be very heavy.

**Comment by mrry** [comment]
@shiyemin Just to clarify, are you fetching or feeding large values in your `sess.run()` calls? If not, it's unlikely that the client to/from master traffic will be heavy... just fetching loss values or summaries. If you are still finding yourself bottlenecked on this, feel free to share your code so we can track it down.

(Note that worker-to-worker traffic is a separate issue, which merits attention.)

**Comment by yaroslavvb** [comment]
@mrry Use-case is training AI agents on OpenAI Universe. You have a stochastic agent like [this one](https://github.com/openai/universe-starter-agent) interacting with an environment external to TensorFlow, so all of the observations to go through `feed_dict`.

**Comment by mrry** [comment]
@yaroslavvb Great, thanks! So, just to be clear, is the overhead in *feeding* (client->worker) or does this model also *fetch* (worker->client) a large amount of data?

**Comment by yaroslavvb** [comment]
Mainly feeding. For Universe you feed images of screen and fetch keyboard commands on each training step, so feeding passes an order of magnitude more data.

Updated the benchmark above to test the feeding direction:
```
python client_transfer_benchmark.py --direction=p2t --profile
1577.86 MB per second
 143.93 MB per second

```

**Comment by yaroslavvb** [comment]
@mrry great, I think fast LocalMaster should cover the main use-cases (since you could always feed to local-master and use worker->worker communication). How do you run the microbenchmark?

**Comment by mrry** [comment]
There are still quite a few redundant copies to remove in the local case, but it's a start :). Here's a link to the benchmark: 

https://github.com/tensorflow/tensorflow/blob/7f65a77b86b15876bc019f85fc7031981e191306/tensorflow/python/client/session_benchmark.py

You should be able to run it directly if you have the PIP package installed. If you have other cases that you'd like to track, feel free to send a PR!

**Comment by yaroslavvb** [comment]
With these changes I'm seeing about 2x improvement in transfer speed:

python client_transfer_benchmark.py --direction=p2t --profile
262.49 MB per second

python client_transfer_benchmark.py --direction=t2p --profile
308.64 MB per second

**Comment by mrry** [comment]
Glad to hear it's moving in the right direction :). Hold on tight because there's another change coming (hopefully in the next push) that should improve feed throughput by a decent amount.

**Comment by yaroslavvb** [comment]
Is bf00bcc the change you had in mind? I don't see any change with that commit on my benchmark, is this new in-memory wrapper activated by default?

__git_version__="b'0.12.1-1591-g4433079'"

python client_transfer_benchmark.py --profile
3203.52 MB per second
280.34 MB per second

**Comment by mrry** [comment]
Most of the improvement from that change will be for the common case where client, master, and worker are in the same address space (like in `session_benchmark.py`). If I've understood the code correctly (since you have a `tf.Session(server.target)`) it should also speed up feeds like this one in Universe's A3C implementation:

https://github.com/openai/universe-starter-agent/blob/ea74a5104ffa19c81562f65000ac56be78e0e0ac/a3c.py#L276

We haven't tried hard (yet) to optimize the case where the client and master are in different processes, although it may be possible to slip in an implementation that avoids a copy when serializing fed tensors into a gRPC call.

**Comment by yaroslavvb** [comment]
Ah, good catch, I was doing out-of-process transfer, updated [client_transfer_benchmark.py](https://gist.github.com/yaroslavvb/59bc8d3d635e1027306486cc418b26aa), now I'm seeing >700MB, nice!

```
python client_transfer_benchmark.py  --profile
(local) 3177.19 MB per second
(grpc) 718.74 MB per second

```

**Comment by yaroslavvb** [comment]
So I tried after compiling latest version (with XLA) and my benchmark got a bit slower

python client_transfer_benchmark.py --direction=p2t --profile
600 MB/s (down from 728)

python client_transfer_benchmark.py --direction=t2p --profile 
1308 MB/s (down from 1879)

**Comment by yaroslavvb** [comment]
@alextp 's work on eliminating numpy copies (ie 5e2cef71 ) could potentially help speed up these local server situations (ie, currently we feed numpy arrays obtained from ALE/Mujoco at each train step, and networks are tiny)

**Comment by skye** [comment]
Ping -- is anyone still looking at this issue or should we close it?

**Comment by mrry** [comment]
After many rounds of optimization, the headline number is certainly no longer true, so let's close this issue for now. I think we've got a gRPC version upgrade pending that should increase the throughput some more. Feel free to open a new issue if you're still running into throughput woes!
